# MID-SEMESTER PROJECT
ARYAH RAO

### CENTRAL IDEA

The goal of our project is to investigate whether senators engage in insider trading. To do this, we are focusing on Senator Nanci Pelosi as our initial point of reference and examining her publicly reported trades in comparison to the fluctuations in stock prices for the same stocks during the relevant time frame.

Once we have imported all the required modules, we can begin working with the datasets, which are all saved as .csv files in our designated 'data' folder.

In [1]:
import os
import os.path
import pandas as pd

datadir = 'data'

#### CONVERTING NANCI'S STOCK TRADING DATA INTO A DATAFRAME USING LoL
Examining the data will provide us with valuable insights into the stock market in general and whether Nanci's conduct had an impact on her capacity to generate profits in a market that is supposed to be free. By observing the fluctuations in the stock prices of the securities that Nanci traded, we can detect any unusual price variations that occurred prior to or following her transactions.

In [2]:
# open csv file
filepath = os.path.join(datadir,'Nanci_history.csv')
with open(filepath, encoding="utf-8-sig") as f:

    # make a LoL
    nanci_LOL = []
    headers = f.readline().strip().split(',')
    
    # iterate lines
    for line in f:
        vals = line.strip().split(',')

        # convert the buy/sell shares into an int
        vals[-1]=int(vals[-1])
        nanci_LOL.append(vals)

# create DataFrame
nanci_data = pd.DataFrame(nanci_LOL, columns=headers)

# fix column name
nanci_data["Date"] = nanci_data["Transaction_Date"]
nanci_data.drop("Transaction_Date", axis=1, inplace=True)

# print DataFrame
nanci_data

Company  Buy_Sell_Shares        Date
0     Nvidia             5000  07-23-2021
1     Nvidia            20000  06-17-2022
2     Nvidia           -25000  07-26-2022
3      Tesla             2500  12-22-2020
4      Tesla             2500  03-17-2022
5      Tesla            -5000  12-20-2022
6  Microsoft            25000  03-19-2021
7  Microsoft             5000  05-24-2022
8  Microsoft           -30000  02-08-2023

From proposal (at all instances): 
Rows: 8
Columns: 3
● Transaction Date - Date of the transaction: STRING
● Symbol - Name or symbol of the stock: STRING
● Buy/Sell Shares - the total number of stocks (shares) bought or sold (positive
means buying stocks and negative means selling stocks): INT

*The CSV file was hand-typed.

#### CONVERTING NASDAQ PRICE CHARTS INTO A DATAFRAME USING DoL

The performance of the NASDAQ stock market serves as a benchmark for evaluating the market trends during the specified time frame.


In [3]:
# open csv file
filepath = os.path.join(datadir,'Nasdaq.csv')
with open(filepath, encoding="utf-8-sig") as f:

    # make a DoL
    nasdaq_DOL = {}  
    headers = f.readline().strip().split(",")

    # create empty lists for headers
    for header in headers:
        nasdaq_DOL[header] = []

    # append to DoL
    for line in f:
        line_list = line.strip().split(",")
        nasdaq_DOL["Date"].append(line_list[0])

        # convert to float
        nasdaq_DOL["Close/Last"].append(float(line_list[1]))
        nasdaq_DOL["Volume"].append(None)
        nasdaq_DOL["Open"].append(float(line_list[3]))
        nasdaq_DOL["High"].append(float(line_list[4]))
        nasdaq_DOL["Low"].append(float(line_list[5]))

# create DataFrame
nasdaq_data = pd.DataFrame(nasdaq_DOL)

# print DataFrame
nasdaq_data

Date  Close/Last Volume      Open      High       Low
0     03/01/2023    11379.48   None  11447.58  11479.00  11349.87
1     02/28/2023    11455.54   None  11451.05  11548.23  11435.39
2     02/27/2023    11466.98   None  11517.19  11565.23  11444.60
3     02/24/2023    11394.94   None  11404.18  11434.36  11334.47
4     02/23/2023    11590.40   None  11636.93  11638.98  11432.58
...          ...         ...    ...       ...       ...       ...
1253  03/08/2018     7427.95   None   7422.77   7435.01   7391.50
1254  03/07/2018     7396.65   None   7311.74   7403.79   7311.74
1255  03/06/2018     7372.01   None   7366.61   7378.03   7319.68
1256  03/05/2018     7330.70   None   7222.89   7350.07   7205.31
1257  03/02/2018     7257.87   None   7099.54   7267.19   7084.83

[1258 rows x 6 columns]

Rows: 1258
Columns: 6
● Date - Date of the stock price: STRING
● Close/Last - Last stock price on the date: FLOAT
● Open - The amount the stock price was at the opening of the market: FLOAT
● High - The highest price the stock has reached that day: FLOAT
● Low - The lowest price the stock has reached that day: FLOAT

https://www.nasdaq.com/market-activity/index/comp/historical

### CONVERTING INDIVIDUAL STOCK PRICE CHARTS INTO DATAFRAMES

Our analysis will involve a comparison of the stock prices of companies to or from which Nanci conducted her trades. In addition, we will compare vital indicators such as trade efficiency and profit/loss to NASDAQ's stock, which will serve as a benchmark for the given period. This will enable us to more accurately determine whether Nanci's trades were profitable or loss-making based solely on market conditions or if there were any cases of insider trading involved.

### Note:
We were required to convert each DataFrame separately using either a List of Dictionaries or a List of Lists, as the data for columns containing stock prices included a '$' symbol. We had to manually avoid the dollar sign and transform the values into integer or float data types.

### TESLA
##### MANUALLY CONVERTING TESLA PRICE CHARTS INTO A DATAFRAME USING DoL.
The behavior of Tesla's stock market throughout the specified timeframe serves as an indicator for evaluating the market trends during that period.


In [4]:
# open csv file
filepath = os.path.join(datadir,'Tesla.csv')
with open(filepath, encoding="utf-8-sig") as f:

    # make a DoL
    tesla_DOL = {}
    headers = f.readline().strip().split(",")

    # create empty lists for headers
    for header in headers:
        tesla_DOL[header] = []

    # append to DoL
    for line in f:
        line_list = line.strip().split(",")
        tesla_DOL["Date"].append(line_list[0])

        # remove '$' and convert to int
        tesla_DOL["Close/Last"].append(float(line_list[1][1:]))
        tesla_DOL["Open"].append(float(line_list[3][1:]))
        tesla_DOL["Volume"].append(int(line_list[2]))
        tesla_DOL["High"].append(float(line_list[4][1:]))
        tesla_DOL["Low"].append(float(line_list[5][1:]))

# create DataFrame
tesla_data = pd.DataFrame(tesla_DOL)

# print DataFrame
tesla_data

Date  Close/Last     Volume      Open      High       Low
0     02/08/2023    201.2900  180673600  196.1000  203.0000  194.3100
1     02/07/2023    196.8100  186010300  196.4300  197.5000  189.5500
2     02/06/2023    194.7600  186188100  193.0100  198.1700  189.9200
3     02/03/2023    189.9800  232662000  183.9500  199.0000  183.6900
4     02/02/2023    188.2700  217448300  187.3250  196.7501  182.6100
...          ...         ...        ...       ...       ...       ...
1254  02/14/2018     21.4873   59221125   21.3893   21.7447   21.2347
1255  02/13/2018     21.5773   68319431   21.0013   21.6127   20.8340
1256  02/12/2018     21.0487   93367138   21.0753   21.2053   20.4167
1257  02/09/2018     20.6947  193866353   21.3287   21.3990   19.6507
1258  02/08/2018     21.0153  154446223   22.8873   23.2413   20.9733

[1259 rows x 6 columns]

Rows: 1259
Columns: 6
● Date - Date of the stock price: STRING
● Close/Last - Last stock price on the date: FLOAT
● Volume - The number of shares traded that day: INT
● Open - The amount the stock price was at the opening of the market: FLOAT
● High - The highest price the stock has reached that day: FLOAT
● Low - The lowest price the stock has reached that day: FLOAT

https://www.nasdaq.com/market-activity/stocks/tsla/historical

### MICROSOFT
##### MANUALLY CONVERTING MICROSOFT PRICE CHARTS INTO A DATAFRAME USING DoL
The behavior of Microsoft's stock market throughout the specified timeframe serves as an indicator for evaluating the market trends during that period.

In [5]:
# open csv file
filepath = os.path.join(datadir,'Microsoft.csv')
with open(filepath, encoding="utf-8-sig") as f:

    # make a DoL
    DOL = {} 
    headers = f.readline().strip().split(",")

    # create empty lists for headers
    for header in headers:
        DOL[header] = []

    # append to DoL
    for line in f:
        line_list = line.strip().split(",")
        DOL["Date"].append(line_list[0])

        # remove '$' and convert to int
        DOL["Close/Last"].append(float(line_list[1][1:]))
        DOL["Open"].append(float(line_list[3][1:]))
        DOL["Volume"].append(int(line_list[2]))
        DOL["High"].append(float(line_list[4][1:]))
        DOL["Low"].append(float(line_list[5][1:]))

# create DataFrame
microsoft_data = pd.DataFrame(DOL)

# print DataFrame
microsoft_data

Date  Close/Last    Volume     Open     High       Low
0     02/08/2023      266.73  54686050  273.200  276.760  266.2100
1     02/07/2023      267.56  50841370  260.530  268.775  260.0800
2     02/06/2023      256.77  22518000  257.440  258.300  254.7800
3     02/03/2023      258.35  29077260  259.542  264.200  257.1000
4     02/02/2023      264.60  39940440  258.820  264.690  257.2500
...          ...         ...       ...      ...      ...       ...
1254  02/14/2018       90.81  34825510   88.510   90.990   88.4100
1255  02/13/2018       89.83  26320030   88.930   90.000   87.8000
1256  02/12/2018       89.13  35657250   88.735   89.780   87.9295
1257  02/09/2018       88.18  63470240   86.300   88.930   83.8300
1258  02/08/2018       85.01  55247060   89.710   89.875   84.7600

[1259 rows x 6 columns]

Rows: 1259
Columns: 6
● Date - Date of the stock price: STRING
● Close/Last - Last stock price on the date: FLOAT
● Volume - The number of shares traded that day: INT
● Open - The amount the stock price was at the opening of the market: FLOAT
● High - The highest price the stock has reached that day: FLOAT
● Low - The lowest price the stock has reached that day: FLOAT

https://www.nasdaq.com/market-activity/stocks/msft/historical

### NVIDIA
##### MANUALLY CONVERTING NVIDIA PRICE CHARTS INTO A DATAFRAME USING DoL
The behavior of Nvidia's stock market throughout the specified timeframe serves as an indicator for evaluating the market trends during that period.


In [6]:
# open csv file
filepath = os.path.join(datadir,'Microsoft.csv')
with open(filepath, encoding="utf-8-sig") as f:

    # make a DoL
    DOL = {}
    headers = f.readline().strip().split(",")

    # create empty lists for headers
    for header in headers:
        DOL[header] = []
    
    # append to DoL
    for line in f:
        line_list = line.strip().split(",")
        DOL["Date"].append(line_list[0])

        # remove '$' and convert to int
        DOL["Close/Last"].append(float(line_list[1][1:]))
        DOL["Open"].append(float(line_list[3][1:]))
        DOL["Volume"].append(int(line_list[2]))
        DOL["High"].append(float(line_list[4][1:]))
        DOL["Low"].append(float(line_list[5][1:]))

# create DataFrame
nvidia_data = pd.DataFrame(DOL)

# print DataFrame
nvidia_data

Date  Close/Last    Volume     Open     High       Low
0     02/08/2023      266.73  54686050  273.200  276.760  266.2100
1     02/07/2023      267.56  50841370  260.530  268.775  260.0800
2     02/06/2023      256.77  22518000  257.440  258.300  254.7800
3     02/03/2023      258.35  29077260  259.542  264.200  257.1000
4     02/02/2023      264.60  39940440  258.820  264.690  257.2500
...          ...         ...       ...      ...      ...       ...
1254  02/14/2018       90.81  34825510   88.510   90.990   88.4100
1255  02/13/2018       89.83  26320030   88.930   90.000   87.8000
1256  02/12/2018       89.13  35657250   88.735   89.780   87.9295
1257  02/09/2018       88.18  63470240   86.300   88.930   83.8300
1258  02/08/2018       85.01  55247060   89.710   89.875   84.7600

[1259 rows x 6 columns]

Rows: 1259
Columns: 6
● Date - Date of the stock price: STRING
● Close/Last - Last stock price on the date: FLOAT
● Volume - The number of shares traded that day: INT
● Open - The amount the stock price was at the opening of the market: FLOAT
● High - The highest price the stock has reached that day: FLOAT
● Low - The lowest price the stock has reached that day: FLOAT

https://www.nasdaq.com/market-activity/stocks/nvda/historical

## TIDYING THE DATA
-> We made an assumption that MICROSOFT's share was sold in the end, so that profit can be calculated.


In [7]:
def split_date(df, delimiter):
    """
    This function splits the Date column in the dataframe into 3 individual columns of day, month 
    and year and returns the new dataframe.
    """
    try:
        df["Day"] = df["Date"].apply(lambda x: int(x.split(delimiter)[1]))
        df["Month"] = df["Date"].apply(lambda x: int(x.split(delimiter)[0]))
        df["Year"] = df["Date"].apply(lambda x: int(x.split(delimiter)[2]))
        df.drop("Date", axis=1, inplace=True)
        return df
    except:
        return df

-> We made a function split_date(df, delimiter) to split Date column into day, month and years.

In [8]:
# splitting date on all dataframes using our helper function
nanci_data = split_date(nanci_data, "-")
nasdaq_data = split_date(nasdaq_data, "/")
tesla_data = split_date(tesla_data, "/")
microsoft_data = split_date(microsoft_data, "/")
nvidia_data = split_date(nvidia_data, "/")

-> Tidying 'nanci_data' by renaming column and setting appropriate columns as index.

In [9]:
# duplicate 'Company' column
nanci_data["Stock"] = nanci_data["Company"]

# remove unnecessary columns
nanci_data.drop(["Company"], axis=1, inplace=True)

# set index
nanci_data.set_index(["Stock", "Year", "Month", "Day"], inplace=True)

# display dataframe
nanci_data

Buy_Sell_Shares
Stock     Year Month Day                 
Nvidia    2021 7     23              5000
          2022 6     17             20000
               7     26            -25000
Tesla     2020 12    22              2500
          2022 3     17              2500
               12    20             -5000
Microsoft 2021 3     19             25000
          2022 5     24              5000
          2023 2     8             -30000

-> Combines all stock information into a dataset, created new column - 'Mean', removed unnecessary columns and set appropriate columns as index.

In [10]:
# creating the dataframe 'stocks' that holds all information together
stocks = pd.concat([nasdaq_data, tesla_data, microsoft_data, nvidia_data], keys=["Nasdaq", "Tesla", "Microsoft", "Nvidia"])

# reset stock and modify to create new index
stocks.reset_index(inplace=True)
stocks["Stock"] = stocks["level_0"]
stocks.set_index(["Stock", "Year", "Month", "Day"], inplace=True)

# Mean value is used as the cal
stocks["Mean"] = (stocks["High"] + stocks["Low"])/2

# remove unnecessary columns
stocks.drop(["level_0", "level_1", "Close/Last", "Open", "High", "Low", "Volume"], axis=1, inplace=True)

# display dataframe
stocks

Mean
Stock  Year Month Day             
Nasdaq 2023 3     1    11414.43500
            2     28   11491.81000
                  27   11504.91500
                  24   11384.41500
                  23   11535.78000
...                            ...
Nvidia 2018 2     14      89.70000
                  13      88.90000
                  12      88.85475
                  9       86.38000
                  8       87.31750

[5035 rows x 1 columns]

-> Combines 'stocks' and 'nanci_data' to create tidy dataset.

In [11]:
# merge stocks and nanci data to create tidy data
merged = stocks.join(nanci_data, how="outer")

### Tidy dataset of all stock trade information


In [12]:
# display the merged and tidy dataset
merged

Mean  Buy_Sell_Shares
Stock     Year Month Day                            
Microsoft 2018 2     8     87.31750              NaN
                     9     86.38000              NaN
                     12    88.85475              NaN
                     13    88.90000              NaN
                     14    89.70000              NaN
...                             ...              ...
Tesla     2023 2     2    189.68005              NaN
                     3    191.34500              NaN
                     6    194.04500              NaN
                     7    193.52500              NaN
                     8    198.65500              NaN

[5035 rows x 2 columns]

Mapping: Independent Variables: ("Stock", "Year", "Month", "Day") --> Dependent Variables: ("Mean", "Buy_Sell_Shares")

In [13]:
# creating a subset with useful rows
merged_subset = merged.dropna(subset=['Buy_Sell_Shares'])

# display dataframe 
merged_subset

Mean  Buy_Sell_Shares
Stock     Year Month Day                            
Microsoft 2021 3     19   230.91180          25000.0
          2022 5     24   257.41500           5000.0
          2023 2     8    271.48500         -30000.0
Nvidia    2021 7     23   288.24500           5000.0
          2022 6     17   247.26500          20000.0
               7     26   254.72500         -25000.0
Tesla     2020 12    22   210.68495           2500.0
          2022 3     17   283.45290           2500.0
               12    20   143.06500          -5000.0

Mapping: Independent Variables: ("Stock", "Year", "Month", "Day") --> Dependent Variables: ("Mean", "Buy_Sell_Shares")

# Computation:
### -> "Did Nanci Pelosi perform insider trading and profit off a free market?" 
- Each of the available metrics is calculated independently on the data in the merged_stocks DataFrame. A set of functions have been implemented to calculate these metrics, and brief explanations of what each function does can be found in the codebase. A more detailed understanding of these functions will be obtained once we implement the metric function. 

#### moneycalculator
The function mentioned here is designed to calculate the total profit or loss made by trades, as well as percentage returns and the cost of doing business. These metrics provide a brief overview of the amount of money involved in the trading venture, and a high profit percentage is a valuable metric for assessing the success of the venture. By analyzing the total profit or loss, we can gain a clear understanding of the financial outcomes of their trades. 

In [14]:
def moneycalculator(df):
    '''
    calculates the total profit/loss made by Nanci through her stock trades.
    
    Returns: 
    make - total profit/loss made
    perc - percentage profit/loss
    cost - total money spent on this venture
    '''
    
    bills=[]
    for x in df['Mean']:
        bills.append(float(x))
    
    i=0
    cost=0
    for y in df['Buy_Sell_Shares']:
        bills[i] = bills[i]*int(y)*(-1)
        
        if bills[i]<0:
            #total cost for buying the shares
            cost+=bills[i]*-1
        i+=1
    
    #since all shares that were bought are sold as well, bills should net 0 if no money was made or lost
    make = 0
    for bill in bills:
        make+=bill
    
    perc = (make/cost)*100

    return make, perc, cost

In [15]:
# unpack result
profit_or_loss, percentage, cost = moneycalculator(merged_subset)

# print result
print("Nanci's net profit is: $", round(profit_or_loss,2))
print("Profit percentage: ", round(percentage, 2), "%")
print("Total money dedicated to trading: $", round(cost, -1))

Nanci's net profit is: $ 546260.38
Profit percentage:  3.72 %
Total money dedicated to trading: $ 14681740.0


##### Helper Functions
These functions are created to help us calculate roi of companies individually first, and then alongside nasdaq in the same time period.

In [16]:
#Helper functions

def company_roe(net_result, company_name):
    '''
    Returns the roi% when net_result(companies as keys and list of integers as values) and company_name is provided
    '''
    profit = sum(net_result[company_name]) # net sun
    cost = sum(num for num in net_result[company_name] if num < 0)*-1 # sum of negative values, and then making them positive

    return profit/cost*100

def nasdaq_company_roe(company_name):
    '''
    Returns the roi% of nasdaq during given time period
    '''

    index = merged_subset.loc[company_name].index[0] # find index
    year, month, day = index # unpack index
    index = ("Nasdaq", year, month, day)
    nasdaq_start = stocks.loc[index, "Mean"] # search nasdaq mean value

    index = merged_subset.loc[company_name].index[2] # find index
    year, month, day = index # unpack index
    index = ("Nasdaq", year, month, day)
    nasdaq_end = stocks.loc[index, "Mean"] # search nasdaq mean value

    return ((nasdaq_end-nasdaq_start)/nasdaq_start)*100

We create a dictionary called 'net_result' to store stock names as keys that hold lists of integers as values corresponding to amount of money used. 
positive --> money added, stock sold
negative --> money subtracted, stock bought

In [17]:
# picks out the amount 
amount = merged_subset['Buy_Sell_Shares'] * merged_subset['Mean'] * -1 # series of amount
amount = amount.tolist() # convert to list

# picks out the net result using the names of the companies
net_result = {}
net_result["Microsoft"] = amount[:3]
net_result["Nvidia"] = amount[3:6]
net_result["Tesla"] = amount[6:]

# display dictionary
net_result

{'Microsoft': [-5772795.0, -1287074.9999999998, 8144550.0],
 'Nvidia': [-1441225.0, -4945300.0, 6368125.0],
 'Tesla': [-526712.375, -708632.25, 715325.0]}

#### effective_ROI
This function calculates the effective return on investment (eROI) by comparing the ROI of the stock to the ROI of the NASDAQ during the same time period. The eROI metric helps determine the profitability of the stock investment in relation to the market conditions. If the eROI significantly exceeds the market performance, it could be an indication of insider trading.

In [18]:
def effective_ROI(net_result):
    '''
    Calculates the return on investment made by Nanci with respect to market conditions that day. 
    It shows if her stock was outperforming the market.
    
    Returns:
    roi - return on investment based on actual values
    eroi - effective return on investment adjusted to NASDAQ values
    mean - mean effective return on investment (+ is always better, - is always bad)
    '''
    # calculates average eROI for all stocks combined
    roi = {}
    eroi = {}

    # calculate company's roi individually
    roi["Microsoft"] = company_roe(net_result, "Microsoft")
    roi["Nvidia"] = company_roe(net_result, "Nvidia")
    roi["Tesla"] = company_roe(net_result, "Tesla")

    # calculate nasdaq's roi during company stock period
    nasdaq_microsoft_roe = nasdaq_company_roe("Microsoft")
    nasdaq_nvidia_roe = nasdaq_company_roe("Nvidia")
    nasdaq_tesla_roe = nasdaq_company_roe("Tesla")

    # calculate company's roi by negaticng nasdaq's
    eroi["Microsoft"] = roi["Microsoft"] - nasdaq_microsoft_roe
    eroi["Nvidia"] = roi["Nvidia"] - nasdaq_nvidia_roe
    eroi["Tesla"] = roi["Tesla"] - nasdaq_tesla_roe

    # calculate mean of company's roi by negaticng nasdaq's
    mean = sum([roi["Microsoft"] - nasdaq_microsoft_roe, roi["Nvidia"] - nasdaq_nvidia_roe, roi["Tesla"] - nasdaq_tesla_roe])/3

    return roi, eroi, mean

A simple function to show what eroi represents.

In [19]:
# helper functions
def eroi_print(eroi):
    if eroi>5:
        print("considered VERY HIGH")
    elif eroi>=1:
        print("considered GOOD")
    elif eroi<1 and eroi>-1:
        print("cconsidered AVERAGE")
    elif eroi<=-1 and eroi>=-5:
        print("considered BAD")
    elif eroi<-5:
        print("considered VERY BAD")

Displaying the result

In [20]:
# unpack result
roi, eroi, mean = effective_ROI(net_result)

# print result
print("Effective Return on Investment: ", round(mean,2), "%")
eroi_print(mean)

Effective Return on Investment:  6.87 %
considered VERY HIGH


##### Helper Function
This function checks whether the eROI and ROI of a stock exceed a certain threshold. By comparing the eROI and ROI metrics to a predefined threshold, the function identifies stocks that may have unusual or unexpected returns. This information can then be used to determine whether insider trading is a potential factor in the stock's performance.

In [21]:
#helper function
def check(eroi, roi, element, threshold):
    '''
    Checks if the eroi and roi of a stock is more than threshold 
    '''
    print("Checking for stock: ", element)
    
    #calculate return
    returns = eroi - roi
    
    # print statements
    
    print("Holding all things in the economy to be constant, the net returns made is: ", round(returns,2), "%")
    print("This can help us conclude: ")

    # threshold possibilities
    if returns>threshold:
        print("High chance of insider trading")
    elif returns<threshold:
        print("Unlikely chance of insider trading")
    elif returns<0:
        print("No chance of insider trading")

#### metric
##### Takes in all the computed data and derives a result that answers our question


This function takes in the eROI metric and checks if it exceeds the market highs. It also considers the other factors calculated in the effective_ROI function and performs operations based on those values. The function assumes that the economy is constant, and therefore, the returns should rarely exceed a certain threshold. The output values from this function provide answers to all of the main questions and provide an overall explanation of the project. Further explanation on the threshold and the market is provided in the comments of the code.

In [22]:
def metric(eroi, roi):
    '''
    Makes a decision on whether Nanci did insider trading or not. It compares the factors to each other and
    using hard-coded market standards. It determines if the trades were unusual or not.
    
    Returns:
    prints sentences which identifies key elements.
    '''
    
    # threshold is the market accepted eROI during the time period (avg return of US financial markets w/o adjusting)
    # since the period was recessionary, 7% threshold that means ~31% yoy returns without adjustting (which is a lot)
    threshold = 7
    print("Threshold value: " ,threshold)
    
    keys = list(eroi.keys())
    for element in keys:
        print()
        # function checks all the data and returns the required dmeographics and output values
        check(eroi[element], roi[element], element, threshold)

In [23]:
#print result
metric(eroi, roi)

Threshold value:  7

Checking for stock:  Microsoft
Holding all things in the economy to be constant, the net returns made is:  8.77 %
This can help us conclude: 
High chance of insider trading

Checking for stock:  Nvidia
Holding all things in the economy to be constant, the net returns made is:  21.32 %
This can help us conclude: 
High chance of insider trading

Checking for stock:  Tesla
Holding all things in the economy to be constant, the net returns made is:  17.54 %
This can help us conclude: 
High chance of insider trading


### CONCLUSION:
Based on the analysis conducted above, it can be concluded that all of the stocks analyzed exhibited signs of potential market manipulation, as the investor was able to generate unreasonably high returns on her investments. Unless the investor is an exceptionally skilled trader, it is likely that she participated in insider trading. The evidence strongly suggests that the investor had access to privileged information that allowed her to make profitable trades with a high degree of accuracy. 